c/yaml2json.c

```diff
@@ -123,50 +123,72 @@ static int yamlReadHandler(void *data, unsigned char *buffer, size_t size, size_
 }
 
 static void decodeParserError(yaml_parser_t *parser, char *errorBuf, size_t errorBufSize) {
+  size_t problemLen = strlen(parser->problem);
+  char problemNative[problemLen + 1];
+  snprintf (problemNative, problemLen + 1, "%s", parser->problem);
+  convertToNative(problemNative, problemLen);
+  char *nativeContext = NULL;
+  if (parser->context) {
+    size_t contextLen = strlen(parser->context);
+    char contextNative[contextLen + 1];
+    snprintf(contextNative, contextLen + 1, "%s", parser->context);
+    convertToNative(contextNative, contextLen);
+    nativeContext = contextNative;
```

Comment:
I've changed to just simply allocating both and freeing at the end. The function decodeParser error just has a bunch of breaks so we can't return early without freeing.

Reply:
You don't need to allocate anything. Simply move that local buffer out of that original `if` scope.

For example, this would cause issues:
```c
char *string = "original text";
if (someCondition) {
  char stringData[9];
  . . . // doing some work
  string = stringData;
} // stringData is not valid after this point and it is a bug to reference it
printf("%s", string); // potential segfault
```
This should be fine:
```c
char *string = "original text";
char stringData[9];
if (someCondition) {
  . . . // doing some work
  string = stringData;
} // stringData is still in the scope where we use it
printf("%s", string);
```

Reply:
Real example:
```c
#include <stdio.h>
#include <string.h>

int main(int argc, char **argv) {

    char *string = "foo";
    {
        char stringData[9];
        printf("stringData address = %p\n", stringData);
        strcpy(stringData, "bar");
        string = stringData;
    }

    char someData[32] = {"you shouldn't see this printrf"};
    printf("someData address   = %p\n", someData);

    printf("result: %s\n", string);

    return 0;
}

> clang -O1 -o main main.c
> ./main

stringData address = 0x16ae1f3f0
someData address   = 0x16ae1f3f0
result: you shouldn't see this printrf

```
And this one is fine:
```c
#include <stdio.h>
#include <string.h>

int main(int argc, char **argv) {

    char *string = "foo";
    char stringData[9];
    {
        printf("stringData address = %p\n", stringData);
        strcpy(stringData, "bar");
        string = stringData;
    }
    
    char someData[32] = {"you shouldn't see this printrf"};
    printf("someData address   = %p\n", someData);

    printf("result: %s\n", string);

    return 0;
}

> clang -O1 -o main main.c
> ./main

stringData address = 0x16d6cb40c
someData address   = 0x16d6cb3e0
result: bar

```

Reply:
Hi Irek. I understand. Thank you for providing an example. :).

I don't know the length of the context until I do a `strlen()`. I was doing the `strlen()` in the null check (condition) that was causing the scoping problem.

Would you prefer I allocate a buffer outside on the stack that should be big enough to contain the context rather than allocate the exact amount on the heap?

`
char contextNative[1024] = {0};
if (parser->context) {
  if (strlen(parser->context < contextNative)) {
    strcpy(contextNative, parser->context);
  }
}
`